examples/hdf5.py

```diff
@@ -4,99 +4,199 @@
+            raise ValueError
+
+        self.x = x
+        self.y = y
+        self.sampleLimit = sampleLimit  # at most 4 times this number of samples will be plotted
+        self.cacheLimit = cacheLimit  # at most 8 times this number of samples will be cached
+        self.cacheSize = 0
+        self.padding = padding
+        self._minCacheLevel = max(2, (len(self) // self.cacheLimit).bit_length())  # n < 2 (ds < 4) has no effect
+        self._maxCacheLevel = (len(self) // self.sampleLimit).bit_length()
+        self._last = None
+        if eager and (self._maxCacheLevel >= self._minCacheLevel):
+            # spawn a thread to eagerly downsample data up to self._maxCacheLevel so
+            # it will be ready for when the user zooms out
+            self._t = threading.Thread(target=lambda: self._loadLevel(self._maxCacheLevel))
+            self._t.start()
+
+    def isRangeValid(self, x1, x2):
+        if self._last is None:
+            return False
+
+        i11, i21, ds1, n = self._last
+        i12, i22, ds2, n = self._range(x1, x2, pad=False)
+        if not (i11 <= i12 <= i22 <= i21):
+            return False
+        i12, i22, ds2, n = self._range(x1, x2, pad=True)
+        return ds1 == ds2
+
+    def __getitem__(self, item):
+        if not isinstance(item, slice) or item.step is not None:
+            raise TypeError
+
+        i1, i2, ds, n = self._last = self._range(item.start, item.stop)
+
+        if n < 2:
+            # downsampling has no effect
+            y = self.y[i1:i2]
+            i = np.arange(i1, i2)
+        else:
+            if n < self._minCacheLevel:
+                y = downsample(self.y, i1, i2, ds)
+            else:
+                if hasattr(self, '_t'):
+                    self._t.join()
+                    del self._t
+                y = self._loadLevel(n)[2 * i1 // ds: 2 * i2 // ds + 1]
+            i = np.arange(i1, i1 + len(y) * ds // 2, ds // 2)
+        print('loaded:', dict(
+            range=[i1, i2],
+            ds=ds,
+            n=n,
+            size=i2-i1,
+            downsampled_size=len(y),
+            cached=n >= self._minCacheLevel
+        ))
+        x = i if self.x is None else self.x[i]
+        return x, y
+
+    def __len__(self):
+        return len(self.y)
+
+    def _range(self, x1, x2, pad=True):
+        if pad:
+            pad = (x2 - x1) * self.padding
+            x1, x2 = x1 - pad, x2 + pad
+        if self.x is None:
+            i1 = max(0, min(len(self), int(x1)))
+            i2 = max(0, min(len(self), int(x2)))
+        else:
+            i1, i2 = np.searchsorted(self.x, [x1, x2]).tolist()
+        ds = max(1, (i2 - i1) // self.sampleLimit)
+        n = ds.bit_length()
+        ds = 2 ** n
+        i1 = (i1 // ds - 1) * ds
+        i2 = (i2 // ds + 1) * ds
+        i1 = max(0, min(len(self), i1))
+        i2 = max(0, min(len(self), i2))
+        return i1, i2, ds, n
+
+    @lru_cache(maxsize=None)
+    def _loadLevel(self, n):
```

Comment:
I don't really understand why this is done in a thread?

Is there something in here that releases the GIL?

Reply:
So the user can explore the plot without it hanging while the cache is being filled.

Reply:
Ahh. Interesting. Still does it need to be a method?

It seems you could just pass in all the inputs here.